pallets/randomness/src/benchmarks.rs

```diff
@@ -39,19 +53,86 @@ benchmarks! {
 	// Thread 'main' panicked at 'set in `set_validation_data`inherent => available before
 	// on_initialize', runtime/moonbase/src/lib.rs:1111
 	// set_babe_randomness_results {}: _(RawOrigin::None)
-	// verify { }
-
-	// // TODO: need to produce Vrf PreDigest using authoring NimbusId and insert both into digests
-	// set_output {
-	// 	// needs to be 2nd block to reflect expected costs every block
-	// 	crate::vrf::set_input::<T>();
-	// 	crate::vrf::set_output::<T>();
-	// 	crate::vrf::set_input::<T>();
-	// }: {
-	// 	crate::vrf::set_output::<T>();
-	// }
 	// verify {}
 
+	// Benchmark for VRF verification and everything else in `set_output`, in `on_initialize`
+	on_initialize {
+		fn decode_32_bytes(input: String) -> [u8; 32] {
+			let output = hex::decode(input).expect("expect to decode input");
+			let mut ret: [u8; 32] = Default::default();
+			ret.copy_from_slice(&output[0..32]);
+			ret
+		}
+		fn decode_key(input: String) -> sr25519::Public {
+			sr25519::Public::unchecked_from(decode_32_bytes(input))
+		}
+		fn decode_pre_digest(input: String) -> PreDigest {
+			let output = hex::decode(input).expect("expect to decode input");
+			const PRE_DIGEST_BYTE_LEN: usize = size_of::<PreDigest>() as usize;
+			let mut ret: [u8; PRE_DIGEST_BYTE_LEN] = [0u8; PRE_DIGEST_BYTE_LEN];
+			ret.copy_from_slice(&output[0..PRE_DIGEST_BYTE_LEN]);
+			Decode::decode(&mut ret.as_slice()).expect("expect to decode predigest")
+		}
+		// Uses values from moonbase alpha storage
+		let raw_nimbus_id = "e0d47c4ea4fb92a510327774bd829d85ec64c06e63b3274587dfa4282f0b8262"
```

Comment:
I think it would be better here if you just take the default Digest and sign its vrf with a random key. Instead of putting the bytes here. is this a possibility?

Reply:
> I think it would be better here if you just take the default Digest and sign its vrf with a random key.

The first commit took this approach, but I could not find a way in `no_std` to generate the schnorrkel::Keypair in-memory and then use it to sign Hash::default() (or anything else). Here is the first commit where I tried this and there is a CI error which shows that the sr25519::Pair struct is not exposed in `no_std`: https://github.com/PureStake/moonbeam/pull/1750/commits/9e850fd5c1603a6844a0577fcf68f7d1c55f5ef4#diff-87d316c2a6be72735791f35cdc983cbc5a9dd8dfdfa94544aad68f7645e42fb9R32-R44

I think the above constraints are designed to prevent signing messages in-memory in `no_std`. The client can do the signature by looking up the key in the keystore and using it to sign, but the method to sign as well as the Private part of the Keypair is not exposed in `no_std` AFAICT.

Reply:
I considered generating a key locally in dev mode, signing Hash::default() == [0u8; 32] and then putting those bytes here instead, but I don't see why that is better than using values from moonbase alpha. I think it's better to benchmark signing a real VrfOutput instead of [0u8; 32].

Reply:
sounds reasonable